### PR TITLE
Use sparse checkout for generate matrix job

### PR DESCRIPTION
**Pull request by benbp**
From a functional standpoint, the generate matrix job only takes a few seconds to run. Depending on the repository, the job can take a total time of 4 minutes before the test jobs are able to start, because azure pipelines does a clone of the entire repository (among other things). This PR makes a change to skip the azure pipelines checkout step, and add a lightweight git clone along with a sparse checkout of just the `eng` and service directory locations (which contain the matrix scripts and/or matrix configs).

This change reduces the total job runtime to around 20 seconds from 4 minutes (repo checkouts + auto-injected policy steps).

### CODE REVIEW

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (alzimmermsft) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:30]: Can `AdditionalParameters.ServiceDirectory` be a list?

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:28]: I tried putting this in a separate script, but of course the pipeline can't load any external files until it does a checkout!

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:30]: At the moment, ServiceDirectory as its passed in through the pipelines is a single value. In the java repo, we've started to add another parameter `TestResourceDirectories` to accommodate a list. My plan is to eventually support a list of directories (for this and ARM templates) across the repositories, so my thinking is when that change comes, we can make the updates here based on what it looks like.

Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:30]: And yes, I think supporting an opt-in to selective checkout for ALL live test jobs is going to be nice to have, but there are more potential edge cases there. I'd like to start investigating that as a follow-on.

Comment (weshaggard) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:30]: I thought you were going to condition this entire block on this parameter. Because we essentially need to revert back to the default full clone if it isn't set. 

Also keep in mind part of the reason we need to fallback to the default is for cases where we are using this on a non-public repository.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:33]: Will this ever be enough? It seems like we would at least need to checkout the directory where the matrix file is located.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:30]: Ah yes, forgot about that case this morning. I'll wrap both steps in a service directory conditional.

Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:30]: Wouldn't a non-public repository still set ServiceDirectory?

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:33]: If there's no service directory, then I assume we're using the default matrix config which is located in `eng`.

**Comment by benbp** [comment]
Pipeline example using this new mode: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=758564&view=logs&j=6865b188-e449-5409-40f9-44b712c50e56&t=48ee3de9-3dc4-5425-2d15-671a2278bddc
Same pipeline but with checkout: false (old behavior): https://dev.azure.com/azure-sdk/internal/_build/results?buildId=758583&view=logs&j=6865b188-e449-5409-40f9-44b712c50e56&t=6865b188-e449-5409-40f9-44b712c50e56

**Comment by mitchdenny** [comment]
Given things aren't on fire right now, I wonder if it is worth taking this opportunity to try and generalize this into a template that takes a list of directories. So that we can do something like this:

```yaml
- template: eng/common/pipelines/templates/steps/sparse-checkout.yml
  parameters:
    Paths:
      - eng/
      - eng/${{parameters.ServiceDirectory}}
```

You're probably already thinking this, but I was thinking you might be able to jump straight to the end-game here at the expense of a few more days of longer job generation (which isn't going to hurt too much).

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (mitchdenny) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:33]: You'll also need to consider auth for the ```-pr``` repositories.

**Comment by benbp** [comment]
@mitchdenny unfortunately I can't use a template, because in order to load templates, we need to have a repository checked out. So it's kind of a chicken or egg problem. I could add a directory parameters argument as a placeholder now that we can extend into going forward?

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:33]: I think the initial approach @weshaggard and I were thinking about is to just do the full checkout for private, since it includes the service connection, but I'll need to find the right place to set `SparseCheckout: false` for that.

**Comment by mitchdenny** [comment]
Templates are evaluated server side. You shouldn't need to checkout in order to use them.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:15]: I think we would need to figure out how to look at the repo name and do a template condition so we don't have to have different code checked into our public and private repos.

**Comment by benbp** [comment]
@mitchdenny  ahh, ok I was confusing script paths with templates. I'll try the latter.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:15]: It's easy enough to do a repo name check to just skip the sparse checkout, but I'd like to figure out a way to load creds for the pr repos in that case.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml]: Updated to include a list or a service directory default.

**Comment by benbp** [comment]
@weshaggard @mitchdenny I updated this to be included as a template, include a list of paths, and to handle private + `-pr` repositories. For `-pr` repos, I had to filter by build definition name, since the repo name variable isn't available at template compile time. If we wanted to do it that way instead, I'd have to add a `condition` to all the steps in the sparse checkout template.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:40]: I'm not sure if it is better to check against the pipeline or repo name? I have a slight preference for the repo name so we aren't as dependent on someone failing to name their pipeline correctly.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:71]: Instead of overriding this here should we pass in a path to the sparse-checkout template and make it the same as the default checkout?

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:40]: Repo name isn't available at template build time :(

Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:40]: https://github.com/Azure/azure-sdk-tools/pull/1452#issuecomment-790977989

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:71]: I'll make it more consistent.

Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml]: Looks like it works fine since everything defaults there anyway. This was a holdover from when I was doing `git clone` into a sub-directory instead of to `.`.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (weshaggard) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:40]: booo... I guess we will have to see how this works out. I don't like how loosely coupled this is.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (mitchdenny) [inline eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:34]: Is there a reason to parameterize this?

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (weshaggard) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:40]: I suspect this is going to come back to bite us once we start using it in more places in the private repos, but for now I guess we can start here and then reconsider other options if we hit an issue.

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/sparse-checkout.yml:10]: It is probably worth adding the repository as a template parameter as I believe this template will be interesting for pipelines we clone multiple pipelines. 

I also suggest adding a template parameter for the working directory for these steps so folks can control where this clone happens.

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/sparse-checkout.yml:19]: The checkout version should also be a template parameter.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:40]: Yeah, I'm not a fan of this approach but can't find an alternative. My thinking is also that the `pr` repos are right now our only standardized name private repos, and people that need to do temporary feature/pre-release testing can just keep `SparseCheckout: false` in their branch until they need to go live.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:34]: The reason I have it this way is I want the `/eng/common/scripts/job-matrix/samples/matrix-test.yml` to be a working sample (partly because non-working samples are bad, and because I use it for testing changes like these). Since I test this sample pipeline in the playground project, it doesn't have access to the 1es pool, hence why I need to override the pool/image.

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (benbp) [reply eng/common/pipelines/templates/steps/sparse-checkout.yml]: I added a repositories object, since with multiple repos we need to pair the working directory with the repo name and commit-ish. Also, I set the default the way it is so that when no repo is specified, we just checkout into the default location, this way any calling template doesn't need to manually specify `WorkingDirectory` when there's only one repo.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (weshaggard) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:40]: It would be nice if they didn't have to have a code commit that would need to be undone once the merge to public. We could consider a pipeline variable that someone could set to opt-in/out of this but lets see if any when this becomes an issue.

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/sparse-checkout.yml:9]: LOL... what is commitish? Why not just CommitSha?

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/sparse-checkout.yml:18]: Won't the clone handle this?

Comment (weshaggard) [reply eng/common/pipelines/templates/steps/sparse-checkout.yml:18]: Oh it doesn't handle it because you are expecting the directory to already exist. We could probably handle this by passing the path to the clone method instead of the `.` but I'm indifferent about which approach to take.

**Review by weshaggard — APPROVED** [review]
A couple comments but otherwise looks reasonable.

File: eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
Comment (benbp) [reply eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml:40]: The variable would make the template harder to define since we'd have to use a condition. I was using a `SparseCheckout` top-level parameter myself for testing, but if I had to check-enable it every time for private repo testing I think I would end up making a commit anyway as part of my workflow.

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (benbp) [reply eng/common/pipelines/templates/steps/sparse-checkout.yml:9]: This is an [official git term](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefcommit-ishacommit-ishalsocommittish), though I knew I'd raise some eyebrows :) Technically it means you can pass a commit or a **branch name**, or any other valid object ref. I'm open to a better known noun that still explains what valid values are available.

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (benbp) [reply eng/common/pipelines/templates/steps/sparse-checkout.yml:18]: The problem with passing a path is then I can't the default in any parent pipeline without having to update every subsequent step with a `workingDirectory` override. I wanted to keep it simple for single-repo scenarios, since devops has the same behavior (single repo clone to `.../1/s`, multi-repo clone to `.../1/s/repo`).

Comment (benbp) [reply eng/common/pipelines/templates/steps/sparse-checkout.yml:18]: Oh I see you mean the default parameter being `.` instead of the variable.

Comment (benbp) [reply eng/common/pipelines/templates/steps/sparse-checkout.yml:18]: The problem with that is we need to construct the path to the cloned repo with expressions we can't use in the scope of the parameter definition default, and so we'd have the path reference inline to the code in multiple places, as opposed to just the one place for directory init and then as `workingDirectory` everywhere else.

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/sparse-checkout.yml:9]: I knew we could pass anything here that resolves to a commit but I think this is the first time I've seen this used. I'm fine keeping it.

File: eng/common/pipelines/templates/steps/sparse-checkout.yml
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/sparse-checkout.yml:18]: I'm good with what you have but I would expect you only need to have it inline in the clone command. All the other git steps can be independent and have the working directory set.

**Pull request merged**
